qi-lib/flow/core/deforest-cps.rkt

```diff
@@ -0,0 +1,323 @@
+#lang racket/base
+
+(provide (for-syntax deforest-pass))
+
+(require (for-syntax racket/base
+                     syntax/parse
+                     "deforest-syntax.rkt"
+                     "../extended/util.rkt"
+                     syntax/srcloc
+                     racket/syntax-srcloc)
+         "deforest-templates.rkt"
+         racket/performance-hint
+         racket/match
+         racket/contract/base)
+
+;; "Composes" higher-order functions inline by directly applying them
+;; to the result of each subsequent application, with the last argument
+;; being passed to the penultimate application as a (single) argument.
+;; This is specialized to our implementation of stream fusion in the
+;; arguments it expects and how it uses them.
+(define-syntax inline-compose1
+  (syntax-rules ()
+    [(_ f) f]
+    [(_ [op f] rest ...) (op f (inline-compose1 rest ...))]))
```

Comment:
This appears to be slightly different from the inline-compose1 introduced in
aef7f88 (minimally incorporate stream fusion into the compiler, 2023-08-18) but
is identical to the fixed version in d215fea (fix stream fusion (from today's qi
meeting), 2023-08-25). If this is copy-pasted, should it be centralized?